### PR TITLE
Python 3 support

### PR DESCRIPTION
**Pull request by timtroendle**
This pull request is big. Let me know if you prefer splitting it up and/or merging upon another branch. There is still some work to do, see [comment](https://github.com/open-research/sumatra/issues/40#issuecomment-102558596) in #40.

Sorry for my trailing whitespace deletion. Should have disabled that before.

### CODE REVIEW

**Comment by ashander** [comment]
Wow (a lot of) great work! To ease review is there a chance you could do
the following?

- Import the sort of todo-list in #40 to a comment here, and clarify what
has and hasn't been done since you posted that comment?
- if there are LOTS of trailing whitespace changes buried in here (I
haven't had a time to look), possibly rebase and use something like
http://stackoverflow.com/questions/3515597/git-add-only-non-whitespace-changes
to modify those commits where lots of whitespace is changed

I'll let @apdavison or @maxalbert speak to the overall merge strategy as
I'm a newcomer to the project.

File: sumatra/datastore/archivingfs.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 from sumatra.core import TIMESTAMP_FORMAT, registry
 
 
-from .base import DataItem, DataKey
```
Comment (apdavison) [inline sumatra/datastore/archivingfs.py]: Unless I'm mistaken, `DataKey` needs to be available in the `archivingfs` namespace.

**Comment by apdavison** [comment]
There aren't so many trailing whitespace deletions, I don't think it's necessary to rebase for that.

**Comment by apdavison** [comment]
This is great work @timtroendle Many thanks. I've reviewed about half the changes so far, I'll try to finish later this week, and I will also test this branch locally (Travis CI doesn't run the system tests at the moment, just the unit tests).

**Comment by ashander** [comment]
Good deal, and you've answered the questions in #40 so I think that
obviates my other suggestion -- sorry for the bother @timtroendle

The only remaining item is 'test extensively'. So, to be helpful I'll try
it this branch locally too

**Comment by timtroendle** [comment]
What's left is the consistent string handling. I'll try that locally and see how many changes that'll be. In case there are quite some changes upcoming, maybe we want to merge this branch first onto another branch than master and open another pull request for that, so that the changes of these new commits are clearer to distinguish.

For testing I'm trying this locally with Python3, Git and Python executables on a Mac (I learned that at least the ``ShelveRecordStore`` has quite some dependency on the system setup). That setup works fine as far as I can tell, but there are so many other setups possible. The more they are covered the better I assume. Python2 might still have problems at the moment due to the unresolved string handling.

**Comment by timtroendle** [comment]
I've added the string handling. To not disturb your review in this pull request, I put these changes on another branch. I opened up a [pull request](https://github.com/timtroendle/sumatra/pull/1) onto this branch here in my own fork. Let me know what you prefer. I can merge right away and you'd have all the changes here.

**Comment by apdavison** [comment]
Yes please, can you merge everything so we have only a single PR to review.

**Comment by timtroendle** [comment]
Done.

**Comment by apdavison** [comment]
I tested this out locally, and the unit tests all pass, but there were some failures in the system tests. To get `test/system/test_ircr.py` to pass, I:

- added `from future import standard_library; standard_library.install_aliases()` to `test_ircr.py`
- changed the file open modes from `rb` to `r` and `wb` to `w` in `test_ircr.py` and `test/system/utils.py`
- added `bytes` to the list of types in `parameters.py`, lines 293 and 337.

I'm not sure the latter is the right solution, though.

**Comment by timtroendle** [comment]
I'll have a look at it. Am I right in the assumption that the only relevant system test is `test/system/test_ircr.py`? From `travis.yaml` this seems to be the only one executed.

**Comment by apdavison** [comment]
The other tests in `test/system/test_ircr.py` require Docker, so can't be run on Travis. You can run them locally, though, if you have Docker (need to build the images first).

**Comment by timtroendle** [comment]
These changes are not enough to make the tests run on my machine unfortunately. Actually there are also failing tests on master -- less though. I'll have to spend more time on this.

If I make this to run I'll merge in the current master as well and fix any potential new problems. My feeling is we should merge this branch soon and make Travis run the tests with Py3, as every new change in the repo might need to be ported as well.

**Comment by timtroendle** [comment]
There is one last test failing that deals with the ShelveRecordStore. I wonder why this test did not fail before -- I thought I did not touch that part since. As no other ``dbm`` module seems to be installed on the Travis instance, it falls back to ``dumbdbm``. Here is a quote from its documentation:

>The dumbdbm module is intended as a last resort fallback for the anydbm module when no more robust module is available. The dumbdbm module is not written for speed and is not nearly as heavily used as the other database modules.

Unfortunately ``dumbdbm`` tries to adjust its behavior to its name and does a ``type(key) == type('') == type('value')`` comparison which fails with unicode. This is fixed in Py3 but remains an issue in Py2. Well we could simply skip that test when there is no _real_ ``dbm`` module. But still, that means that running Sumatra with a ``ShelveRecordStore`` will crash on such machines.

Have to think about that tomorrow again. Any opinion from your sides on this?

**Comment by apdavison** [comment]
> Unfortunately dumbdbm tries to adjust its behavior to its name and does a type(key) == type('') == type('value') comparison which fails with unicode. This is fixed in Py3 but remains an issue in Py2. Well we could simply skip that test when there is no real dbm module. But still, that means that running Sumatra with a ShelveRecordStore will crash on such machines.

This sounds like a Python bug. Is there an issue on the Python bug tracker for this? Anyway, if it is a Python bug then _any_ use of `shelve` or `pickle` with unicode will fail on such machines, so it's not a Sumatra-specific issue. Skipping the test in such situations seems like the right approach.

**Comment by apdavison** [comment]
> I wonder why this test did not fail before

This may be because I switched from using the system Python distribution on the Travis boxes to Miniconda (in order to be able to install recent versions of NumPy, etc. without compilation).

**Comment by timtroendle** [comment]
Ok, that took a while but I think I have it. The reason why only this one line failed was line 26 in `shelve_store.py` in which the Unicode string is transformed into a Py2 `str`. In the test, the shelve store is used directly though. I suppose this is not a use-case and hence the `ShelveRecordStore` should work on systems without `ndbm`. 

>This sounds like a Python bug. Is there an issue on the Python bug tracker for this? Anyway, if it is a Python bug then any use of shelve or pickle with unicode will fail on such machines, so it's not a Sumatra-specific issue. Skipping the test in such situations seems like the right approach. 

It's not actually a bug, it just don't accept Unicode. With line 26 in `shelve_store.py` I think this is solved for Sumatra, isn't it?

> This may be because I switched from using the system Python distribution on the Travis boxes to Miniconda (in order to be able to install recent versions of NumPy, etc. without compilation).

Right, I can reproduce that on my Linux machine.

**Comment by timtroendle** [comment]
My last commit was rather radical as I removed the whole unused and currently failing test setup of `test_web`. Let me know if that was too radical and you'd prefer keeping it for later.

**Comment by apdavison** [comment]
It's ok to remove it in the interest of getting this PR merged, since it is not being used at the moment. I can always retrieve it from the repository history when I get round to reimplementing tests of the web browser interface.

**Comment by timtroendle** [comment]
Great! Glad to see that this worked out. I hope this will be of use for some people so that it was worth the effort ... ;-)

By the way, a good reference for continuing to keep the code base Py2/Py3 compatible can be found at the [python-future project](http://python-future.org/compatible_idioms.html).

**Comment by ashander** [comment]
Thanks Tim!



- Jaime

**Pull request merged**